src/platforms/posix/px4_daemon/server.cpp

```diff
@@ -0,0 +1,299 @@
+		return;
+	}
+
+	// We open the client's specific pipe to write the return value and stdout back to.
+	// The pipe's path is created knowing the UUID of the client.
+	char path[RECV_PIPE_PATH_LEN] = {};
+	int ret = get_client_recv_pipe_path(packet.header.client_uuid, path, RECV_PIPE_PATH_LEN);
+
+	if (ret < 0) {
+		PX4_ERR("failed to assemble path");
+		return;
+	}
+
+	int pipe_fd = open(path, O_WRONLY);
+
+	if (pipe_fd < 0) {
+		PX4_ERR("pipe open fail");
+		return;
+	}
+
+	// To execute a command we start a new thread.
+	pthread_t new_pthread;
+
+	// We need to copy everything that the new thread needs because we will go
+	// out of scope.
+	RunCmdArgs *args = new RunCmdArgs;
+	memcpy(args->cmd, packet.payload.execute_msg.cmd, sizeof(args->cmd));
+	args->client_uuid = packet.header.client_uuid;
+	args->pipe_fd = pipe_fd;
+	args->is_atty = packet.payload.execute_msg.is_atty;
+
+	if (0 != pthread_create(&new_pthread, NULL, Server::_run_cmd, (void *)args)) {
+		PX4_ERR("could not start pthread");
+		delete args;
+		return;
+	}
+
+	// We keep two maps for cleanup if a thread is finished or killed.
+	_client_uuid_to_pthread.insert(std::pair<uint64_t, pthread_t>
+				       (packet.header.client_uuid, new_pthread));
+	_pthread_to_pipe_fd.insert(std::pair<pthread_t, int>
+				   (new_pthread, pipe_fd));
+}
+
+
+void
+Server::_kill_cmd_packet(const client_send_packet_s &packet)
+{
+	// TODO: we currently ignore the signal type.
+
+	pthread_t pthread_to_kill = _client_uuid_to_pthread[packet.header.client_uuid];
+
+	int ret = pthread_cancel(pthread_to_kill);
+
+	if (ret != 0) {
+		PX4_ERR("failed to cancel thread");
+	}
+
+	_cleanup_thread(packet.header.client_uuid);
+
+	// We don't send retval when we get killed.
+	// The client knows this and just exits without confirmation.
+}
+
+
+
+void
+*Server::_run_cmd(void *arg)
+{
+	RunCmdArgs *args = (RunCmdArgs *)arg;
+
+	// Copy arguments so that we can cleanup the arg structure.
+	uint64_t client_uuid = args->client_uuid;
+	int pipe_fd = args->pipe_fd;
+	bool is_atty = args->is_atty;
+	std::string message_str(args->cmd);
+
+	// Clean up the args from the heap in case the thread gets canceled
+	// from outside.
+	delete args;
+
+	// We register thread specific data. This is used for PX4_INFO (etc.) log calls.
+	CmdThreadSpecificData *thread_data_ptr;
+
+	if ((thread_data_ptr = (CmdThreadSpecificData *)pthread_getspecific(_instance->_key)) == NULL) {
+		thread_data_ptr = new CmdThreadSpecificData;
+		thread_data_ptr->pipe_fd = pipe_fd;
+		thread_data_ptr->is_atty = is_atty;
+		thread_data_ptr->packet.header.msg_id = client_recv_packet_s::message_header_s::e_msg_id::STDOUT;
+
+		(void)pthread_setspecific(_instance->_key, (void *)thread_data_ptr);
+	}
+
+	// Run the actual command.
+	int retval = Pxh::process_line(message_str, false);
+
+	// Report return value.
+	_send_retval(pipe_fd, retval, client_uuid);
+
+	// Clean up before returning.
+	_instance->_cleanup_thread(client_uuid);
```

Comment:
yes, that's probably the simplest way. or signal the server thread to remove the uuid, but this also requires a queue with synchronization.